### PR TITLE
fix: 11636 VirtualHasher performance improvements

### LINKED ISSUE #11636: Reduce object allocations in the new virtual hasher
This is a follow-up for https://github.com/hashgraph/hedera-services/issues/10961.

A few optimizations in the new `VirtualHasher` implementation that may help with Java object allocations:

1. When a task is created for a leaf node, there is no need to create its `ins` array. It will not be used anyway. Either a null or a real leaf will be used as a task input

2. Sometimes a sibling task is created only to call `setParent()` and `setData(null)` on it. Instead, a null input may be set on the parent task, bypassing creating a task object

Another issue is not directly related to object allocations, but still looks like a valuable optimization:

3. There is a loop to process existing tasks starting from `stack[curRank]` to `curPath`. After the loop, `stack` for the current rank is invalidated, i.e. set to `INVALID_PATH`. However, it may happen that `curPath` is actually in the same chunk as `stack[curRank]`. In this case, all existing tasks between `curPath` and the end of the chunk will hang in the tasks map and will be processed after the last leaf only (in the loop to set null data for all tasks remaining in the map), despite these tasks are known to be clear. To fix it, `stack[curRank]` should be set to `curPath - 1`, if `curPath` is in the same rank

### PR DESCRIPTION
**Pull request by imalygin**
VirtualHasher performance improvements:

- updated chunk height to 3
- introduced `complete` method to `ChunkHashTask` to handle clean nodes optimally

**Description**:
VirtualHasher performance improvements:

- updated chunk height to 3
- introduced `complete` method to `ChunkHashTask` to handle clean nodes optimally

**Related issue(s)**:

Fixes #11636

### CODE REVIEW

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java [hunk 1]
```diff
@@ -145,51 +151,58 @@ public Hash hash(
...
+            super(pool, dependencies);
+            ins = numHashes > 0 ? new Hash[numHashes] : null;
+        }
+
+        @Override
+        protected boolean exec() {
+            return true;
+        }
+
+        void setHash(int index, Hash hash) {
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java:169]: Please, make all args final

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java [hunk 2]
```diff
@@ -334,7 +345,10 @@ public Hash hash(
         // the root task below. When the root task is done executing, that is it produced
         // a root hash, this hash is set as an input dependency for this result task, where
         // it's read and returned in the end of this method
-        ChunkHashTask resultTask = new ChunkHashTask(HASHING_POOL, INVALID_PATH, 1);
+        /**
+         * A task holding the resulting hash. Used to synchronize all parallel computations.
+         */
+        HashHoldingTask resultTask = new HashHoldingTask(HASHING_POOL, 1, 1);
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java:351]: final

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java (hunk 1 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java]: Done

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java (hunk 2 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java]: Done

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java [hunk 3]
```diff
@@ -400,10 +409,21 @@ public Hash hash(
                     while (curStackPath < Math.min(curPath, lastPathInCurStackChunk)) {
                         final ChunkHashTask t = map.remove(curStackPath);
                         assert t != null;
-                        t.setData(null);
+                        t.complete();
                         curStackPath++;
                     }
-                    stack[curRank] = INVALID_PATH;
+                    /*
+                       It may happen that curPath is actually in the same chunk as lastPathInCurStackChunk.
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java:416]: "... in the same chunk as stack[curRank]"

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java (hunk 3 above)
Comment (imalygin) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java]: Done

**Review by alittley — APPROVED** [review]
Approved WRT hashgraph owned files

**Review by artemananiev — DISMISSED** [review]

**Review by alittley — APPROVED** [review]

**Comment by OlegMazurov** [comment]
`VirtualMapBench` result in GCP:
```
Benchmark                             (keySize)   (maxKey)  (numFiles)  (numRecords)  (numThreads)  (recordSize)  Mode  Cnt      Score   Error  Units
CryptoBenchMerkleDb.transferPrefetch         24  500000000        6000        100000            32          1024  avgt       17921.621           s/op
```
Baseline score: 19066.948 (lower is better).

**Review by OlegMazurov — APPROVED** [review]

**Pull request merged**